com.unity.netcode.gameobjects/Tests/Runtime/NetworkBehaviourUpdaterTests.cs

```diff
@@ -371,30 +336,10 @@ private IEnumerator StartClientsAndServer(bool useHost, int numberOfClients, Gam
                     }
                 }
 
-                var allClientsCompleted = false;
-                AdvanceTimeOutPeriod();
-
-                // Wait until all clients have had their NetworkVariables updated
-                while (!allClientsCompleted && !HasTimedOut())
-                {
-                    var completedCount = 0;
-                    foreach (var clientNetVarContainer in clientSideNetVarContainers)
-                    {
-                        if (clientNetVarContainer.HaveAllValuesChanged(NetVarValueToSet))
-                        {
-                            completedCount++;
-                        }
-                    }
-
-                    allClientsCompleted = completedCount == clientSideNetVarContainers.Count;
-
-                    yield return new WaitForSeconds(tickInterval);
-                }
-
-                Assert.True(!HasTimedOut(), $"Timed out waiting for client side NetVarContainers to report all NetworkVariables have been updated!");
-
-                // This really should never fail as it should timeout first
-                Assert.True(allClientsCompleted, "Not all client side NetworkVariables reported they were updated with target value!");
+                yield return WaitForConditionOrTimeOut((c) =>
```

Comment:
Something to consider here: What this test is really trying to do is wait for messages to be sent and received and then assert that the values have changed. Waiting for the values to change like this is kind of mixing the precondition with the assertion - if the assertion fails we wait forever.

In general, I would like to see us moving in the direction in our test code of waiting for distinct preconditions that are not part of what's being tested, and then asserting on the expected result. So in this case, using `WaitForMessageOfType<SnapshotDataMessage>` or `WaitForMessageOfType<NetworkVariableDeltaMessage>` (not sure if snapshots are on or not for this test) and then asserting that all the values have changed.

Reading this test as it's written now, if I were to change it to a "given-when-then" statement, I would word it as "Given an object with network variables on multiple clients, when all clients have changed the values of their variables, waiting for them to change doesn't time out".

I know this is a pretty pedantic comment, but I think moving toward smaller, more discreet tests that test only one thing and are very clear about what they're testing by what's inside the one assert, will ultimately make our test code better, cleaner, and easier to maintain.

Reply:
There are other places in this PR where this comment also applies, but I'm going to only mention it the once. It might be something to address later rather than now.

Reply:
(This comment, btw, is directed at myself as well. I've been guilty of this in the past, too. I've been trying to learn more about good test design and this is something I've learned has value over the last few months.)

Reply:
Actually, that is why I switched it over to the +new+ WaitForConditionOrTimeOut.  It will wait for a condition to be reached and if the condition is never met it will stop waiting and the new assert below checks to see if it timed out or not.  If not, it moves on, but if so then it asserts with a meaningful/unique message that makes troubleshooting a failed integration test easier to find.